### DISCUSSION TITLE
New feature ideas

### DISCUSSION THREAD

**Discussion by Joly0** [hollowaykeanho/Upscaler#30] [Ideas]
Hey, i have thought about some new features for this project and would like to have some feedback.

The features i had in mind would be these:

- Auto install ffmpeg if not installed (would work good for windows, not sure about linux)
- Automatically check for updates/new relases
- Docker version

**Comment by hollowaykeanho** [comment]
> Auto install ffmpeg if not installed (would work good for windows, not sure about linux)

On linux side there are various steps depending on distro. Normally I wouldn't want to pull in the steps since Linux supports various CPU that Windows can't reach (e.g. your router's OS).

Are you referring to building Windows Store's installer?

> Automatically check for updates/new relases

This should be fine. Quite easy in fact. =) File an Issue ticket for this and we can work on it.

> Docker version

There was another request on my M'sia side as well but I can't visualize where to use it and how to use it. You got any use case?

I only use docker for k9s server applications. I don't use docker outside that scope.

----

I'm thinking on implementing the parallel processing using score approach (e.g. 2, 3, 4, etc) to further improve the performance once I figure out the mutex locking mechanism for both shells.

**Reply by Joly0** [reply]
> 
> On linux side there are various steps depending on distro. Normally I wouldn't want to pull in the steps since Linux supports various CPU that Windows can't reach (e.g. your router's OS).
> 
> Are you referring to building Windows Store's installer?
> 
I thought about using winget or atleast its package repository to automatically install ffmpeg. Its quite simple with winget using "winget install ffmpeg" command, though winget is really only pre-installed in windows 11 21h1. But i might still be able to build a small downloader for it to search for the latest release of ffmpeg in the winget package repo and download and install it that way.

I have no idea on the linux side, it might be quite difficult to do that, as there are different distros, with different package managers like apt, apk, yum and others, but i still think it might be helpful for new users

> This should be fine. Quite easy in fact. =) File an Issue ticket for this and we can work on it.
> 
Alright, will do that

> There was another request on my M'sia side as well but I can't visualize where to use it and how to use it. You got any use case?
> 
> I only use docker for k9s server applications. I don't use docker outside that scope.
> 
I thoughr about two implementations. Either have a third party file watcher that watches a specified folder for a media file and a config file and would start upscaler using the settings of that config file. That would mean this could be run without stopping and starting the container to upscale a new file.
The other method would be to tell a script to look at container start for a media file plus a config file in a specified folder and then start upscale with the settings.
In the second scenario, the configs could be added as environment variables aswell, but i thought this might make it too hard, i would prefer the first method.
There are various projects that do such things for re-encoding video files aswell for example to save space, and i thought this could be done for this project aswell.

> 
> I'm thinking on implementing the parallel processing using score approach (e.g. 2, 3, 4, etc) to further improve the performance once I figure out the mutex locking mechanism for both shells. 
> 
That sounds quite interesting. Could you explain it a bit deeper on how this works and how you would implement it?

**Comment by hollowaykeanho** [comment]
> I thought about using winget or atleast its package repository to automatically install ffmpeg. Its quite simple with winget using "winget install ffmpeg" command, though winget is really only pre-installed in windows 11 21h1. But i might still be able to build a small downloader for it to search for the latest release of ffmpeg in the winget package repo and download and install it that way.

Sounds good to me (I had a "fun" time messing with Jean's benchmark recently because of dependencies management). I think what we can do is agree on an argument (e.g. `--setup`?) for calling and then on powershell you can implement the `winget`. On linux side what I can do is checking all the required binary and present instructions to user.

> That sounds quite interesting. Could you explain it a bit deeper on how this works and how you would implement it?

Basically the score number hard limits how many upscaling process can be run at each instance (score 2 means 2 at a time). These Shell scripts don't offer any good concurrent sync mechnism (mutex / channel). I'm aware POSIX shell can do directory sync: https://unix.stackexchange.com/questions/401415/mutex-and-semaphore-like-in-shell-script because `mkdir` check and create at the same time. Does powershell has any creation command that do "check and create" simultenously?

The algorithm change from:

```
1. Read control state
2. extract 1 frame
3. upscale
4. repeat (1)
```

To:

```
1. Skip this step if extraction state is done or otherwise extract all frames.
2. Save extraction state. 
3. Skip this step if upscaled state is done or otherwise loop:
  3.1. scan sync output control file/directory for status.
  3.2. designate remaining upscale job to thread.
  3.3. repeat 3.1 for remaining frames.
4. Save upscaled state.
```

**Comment by hollowaykeanho** [comment]
Note that when we bump into using proper language, it signifies we had pushed the shell scripts to the limit already.

To go beyond this phase is to study the NCNN and convert entire thing into a proper source codes. This can take time.

**Comment by hollowaykeanho** [comment]
> what we can do is agree on an argument (e.g. --setup?)

Good for you (or do you have other creative names)? I'll open an issue ticket.
